agent/consul/acl.go

```diff
@@ -445,68 +378,37 @@ func (r *ACLResolver) resolveIdentityFromToken(token string) (structs.ACLIdentit
 	metrics.IncrCounter([]string{"acl", "token", "cache_miss"}, 1)
 
 	// Background a RPC request and wait on it if we must
-	var waitChan chan *remoteACLIdentityResult
+	var waitChan <-chan singleflight.Result
 
 	waitForResult := cacheEntry == nil || r.config.ACLDownPolicy != "async-cache"
 
-	r.asyncIdentityResultsMutex.Lock()
-	// check if resolution of this token is already ongoing
-	waiters, ok := r.asyncIdentityResults[token]
-	if !ok || waiters == nil {
-		// only initialize the slice of waiters if need be (when this token resolution isn't ongoing)
-		waiters = make([]chan *remoteACLIdentityResult, 0)
-	}
-	if waitForResult {
-		// create the waitChan only if we are going to block waiting
-		// for the response and then append it to the list of waiters
-		// Because we will block (not select or discard this chan) we
-		// do not need to create it as buffered
-		waitChan = make(chan *remoteACLIdentityResult)
-		r.asyncIdentityResults[token] = append(waiters, waitChan)
-	}
-	r.asyncIdentityResultsMutex.Unlock()
-
-	if !ok {
-		// only start the RPC if one isn't in flight
-		go r.resolveIdentityFromTokenAsync(token, cacheEntry)
-	}
+	waitChan = r.identityGroup.DoChan(token, func() (interface{}, error) {
+		identity, err := r.fetchAndCacheIdentityFromToken(token, cacheEntry)
+		return identity, err
+	})
 
 	if !waitForResult {
 		// waitForResult being false requires the cacheEntry to not be nil
+		// TODO(rb): should this return (nil, acl.ErrNotFound) when the identity is nil?
```

Comment:
I think that would be nice although things are going to nil check the Authorizer anyways so with the existing code its probably not strictly necessary.

Reply:
Ok. I only noticed it when comparing `identity` vs `legacy` and noticed very slight inconsistencies.